### PR TITLE
allow browsing feature list in feature mode

### PR DESCRIPTION
**Pull request by 3nids**
This is somewhat a port of [ItemBrowser](https://plugins.qgis.org/plugins/itembrowser/) plugin.

image

* arrows to browse selection
* flash / pan / zoom checkable buttons
* progress count

### CODE REVIEW

File: src/gui/attributetable/qgsdualview.h [hunk 1]
```diff
@@ -65,9 +65,19 @@ class GUI_EXPORT QgsDualView : public QStackedWidget, private Ui::QgsDualViewBas
...
       AttributeEditor = 1
     };
-
     Q_ENUM( ViewMode )
 
+
+    //! Action on the map canvas when browsing the list of features
+    enum FeatureListBrowsingAction
+    {
+      NoAction = 0,
```
Comment (nyalldawson) [inline src/gui/attributetable/qgsdualview.h]: Dox

**Comment by nyalldawson** [comment]
Code looks good to me!

From a UX perspective I wonder if this new widget would fit better at the bottom of the list instead of above it. @nirvn?

Have you considered adding global shortcuts for these actions?

**Comment by nirvn** [comment]
@3nids , OK, first things first: yay, this is really cool and handy :smile: 

Big picture comment: IMHO, we shouldn't limit this nice way to navigate features to the dual view mode, we should have it for both the table view and the dual view.

That big picture comment said, UI wise, I have the same feeling as @nyalldawson here, we have to move it around a bit (especially if you agree with the above comment). Here's my poorly drawn recommendation:
image

Other comments:
- IMHO the [ < ] and the [ > ] buttons should have the flat style turned on (i.e. borderless when not hovered)
- For the flash / pan / zoom buttons, IMHO I think we need to make it clearer those are _toggles_, not individual push buttons. I'd vote for bordeless unless selected (just like the bottom right view toggle buttons)
- A [ << ] and [ >> ] pair of buttons to jump to first / last features could be useful too.

File: src/gui/attributetable/qgsdualview.cpp [hunk 2]
```diff
@@ -401,13 +418,74 @@ void QgsDualView::insertRecentlyUsedDisplayExpression( const QString &expression
...
 
-void QgsDualView::mFeatureList_aboutToChangeEditSelection( bool &ok )
+void QgsDualView::updateEditSelectionProgress( int progress, int count )
+{
+  mProgressCount->setText( QStringLiteral( "%1 / %2" ).arg( progress + 1 ).arg( count ) );
+  mPreviousFeatureButton->setEnabled( progress > 0 );
+  mNextFeatureButton->setEnabled( progress + 1 < count );
+}
+
+void QgsDualView::panOrZoomToFeature( QgsFeatureIds featureset )
```
Comment (m-kuhn) [inline src/gui/attributetable/qgsdualview.cpp]:
```suggestion
void QgsDualView::panOrZoomToFeature( const QgsFeatureIds &featureset )
```

File: src/gui/qgsmapcanvas.h [hunk 3]
```diff
@@ -253,9 +253,11 @@ class GUI_EXPORT QgsMapCanvas : public QGraphicsView
 
     /**
      * Centers canvas extent to feature ids
-        \param layer the vector layer
-        \param ids the feature ids*/
-    void panToFeatureIds( QgsVectorLayer *layer, const QgsFeatureIds &ids );
+     * \param layer the vector layer
+     * \param ids the feature ids
+     * \param alwaysRecenter if false, the canvas is only recenter if the bounding box is not contained within the current extent
```
Comment (m-kuhn) [inline src/gui/qgsmapcanvas.h]:
```suggestion
     * \param alwaysRecenter if false, the canvas is only recentered if the bounding box is not contained within the current extent
```

**Comment by 3nids** [comment]
> Have you considered adding global shortcuts for these actions?

Not really at first. Any idea of possible keys?

> Big picture comment: IMHO, we shouldn't limit this nice way to navigate features to the dual view mode, we should have it for both the table view and the dual view.

We have thought about it, but the main issue is that there is no concept of currently edited feature in the attribute table. We would either have to play with the selection, or adding a concept of current feature in the table view, which in both case are confusing.
In the case of the attribute table, we could have an auto pan/auto scale only (no browsing or progress count) that would work well with the selection.

> That big picture comment said, UI wise, I have the same feeling as @nyalldawson here, we have to move it around a bit (especially if you agree with the above comment)

Since my disagreement ;), I believe it is better placed above (or under) the list, so it has the exact same  width as the list. But I can move it under it if you prefer? 

I will try to polish the look of the buttons.

**Comment by 3nids** [comment]
updated buttons 
image

and with first/last buttons
image

**Comment by DelazJ** [comment]
missing tag for docs...

**Comment by nirvn** [comment]
Much better look, thanks. 

I see the issue with the table view, it'd indeed require a fair share of extra work.

On the placement front, I'm still not a big fan of it being on top. I think moving it under the feature list as you suggested is a good compromise.

**Comment by 3nids** [comment]
@nirvn here we go 
image

@DelazJ rebased with feature tag and more descriptive commit

**Comment by nirvn** [comment]
What's up with os x and black text on dark button (see expression button and the filter button)?

**Comment by nirvn** [comment]
Re your relocated widget, very nice! 

I'm still tickled by the label in the middle that's center justified by not centered due to toggle buttons.

What about going with an alternative layout, where you'd regroup all nav buttons together, followed by label, then toggle buttons right justified. Ie

[<<][<][>][>>] 1/10  ........... [1][2][3]

?

I can see the prev and next button located next to each other as good UX.

**Comment by 3nids** [comment]
@nirvn for the expression widget I had fixed the issue on mac in 0a10dc504dec344ce15933c6aaaa31128140dccd but apparently you broke it in 89693606bf4ea042791a71da7bde7025af0cb65f

I reorganized the buttons as suggested and it's indeed better. I also reduced the font size of the label a bit.
image

**Comment by nirvn** [comment]
@3nids , that commit is for the expression combo box widget, AFAIK not related to the buttons seen in your screenshot. 

The reordered buttons are looking great.

**Comment by nyalldawson** [comment]
@nirvn @3nids 

testing this out -- isn't the "flash" setting independant of the zoom/pan setting? If I click the zoom button, flash becomes unchecked, but the features still flash when navigating through the table...

Also - we should make the label a text edit instead, and allow users to enter a number to jump directly to a feature

**Comment by 3nids** [comment]
@nyalldawson in the current implementation, flashing is indeed in every mode. In the highlighting PR (#9992), I switched that highlighting is optional. If we stick to flashing for efficiency reason, I'll bring the new way.

not sure about the line edit, it might make the UI a bit heavy now?

**Comment by 3nids** [comment]
@nyalldawson the requested behavior (flashing independent from panning/zooming) has been implemented for flashing only (not identify)

**Pull request merged**